services/services.go

```diff
@@ -109,111 +125,52 @@ type Advertisement interface {
 	Advertise(ad Advertisement) error
 
 	// Unadvertise indicates a given instance is no longer available
-	Unadvertise(service, id string) error
+	Unadvertise(service ServiceQuery, id string) error
 
 	// Query returns metadata and a list of available instances for a given service
-	Query(service string, opts QueryOptions) (Service, error)
+	Query(service ServiceQuery, opts QueryOptions) (Service, error)
 
 	// Watch returns a watch on metadata and a list of available instances for a given service
-	Watch(service string, opts QueryOptions) (xwatch.Watch, error)
-}
-
-type service struct {
-	instances   []ServiceInstance
-	replication ServiceReplication
-	sharding    ServiceSharding
-}
-
-func (s *service) Instance(instanceID string) (ServiceInstance, error) {
-	for _, instance := range s.instances {
-		if instance.ID() == instanceID {
-			return instance, nil
-		}
-	}
-	return nil, errInstanceNotFound
-}
-func (s *service) Instances() []ServiceInstance                 { return s.instances }
-func (s *service) Replication() ServiceReplication              { return s.replication }
-func (s *service) Sharding() ServiceSharding                    { return s.sharding }
-func (s *service) SetInstances(insts []ServiceInstance) Service { s.instances = insts; return s }
-func (s *service) SetReplication(r ServiceReplication) Service  { s.replication = r; return s }
-func (s *service) SetSharding(ss ServiceSharding) Service       { s.sharding = ss; return s }
-
-type serviceReplication struct {
-	replicas int
+	Watch(service ServiceQuery, opts QueryOptions) (xwatch.Watch, error)
 }
 
-func (r *serviceReplication) Replicas() int                          { return r.replicas }
-func (r *serviceReplication) SetReplicas(rep int) ServiceReplication { r.replicas = rep; return r }
-
-type serviceSharding struct {
-	numShards int
-}
-
-func (s *serviceSharding) NumShards() int                     { return s.numShards }
-func (s *serviceSharding) SetNumShards(n int) ServiceSharding { s.numShards = n; return s }
-
-type serviceInstance struct {
-	id       string
-	service  string
-	zone     string
-	endpoint string
-	shards   shard.Shards
+// PlacementService handles the placement related operations for registered services
+// all write or update operations will persist the generated snapshot before returning success
+type PlacementService interface {
+	BuildInitialPlacement(service ServiceQuery, hosts []PlacementInstance, shardLen int, rf int) (PlacementSnapshot, error)
+	AddReplica(service ServiceQuery) (PlacementSnapshot, error)
+	AddHost(service ServiceQuery, candidateHosts []PlacementInstance) (PlacementSnapshot, error)
+	RemoveHost(service ServiceQuery, host PlacementInstance) (PlacementSnapshot, error)
+	ReplaceHost(service ServiceQuery, leavingHost PlacementInstance, candidateHosts []PlacementInstance) (PlacementSnapshot, error)
+
+	// Snapshot gets the persisted snapshot for service
+	Snapshot(service ServiceQuery) (PlacementSnapshot, error)
```

Comment:
As discussed, would be good to have this `PlacementService` operate under the context of a given service already.